### PR TITLE
Error handling (ChakraCore)

### PR DESCRIPTION
**Pull request by jasongin**
I copied the headers from the V8 branch, then added error checking to all the JSRT API calls. This was a bit simpler than the V8 error-handling implementation, since the JSRT APIs already use basically the same error handling and try-catch semantics.

Most of the `addons-abi` tests pass, with the exception of known failures due to [NAPI string API bugs](https://github.com/nodejs/abi-stable-node/issues/63). However to avoid having lots of already-reviewed changes in this code review I'm not updating the tests in the PR; I'll copy them over separately.

**WARNING:** This PR is built on top of the [property accessors PR](https://github.com/nodejs/abi-stable-node/pull/68) that hasn't been merged yet. Please review that PR first, or just look at the second of the two commits here.

### CODE REVIEW

**Comment by jasongin** [comment]
Fixes https://github.com/nodejs/abi-stable-node/issues/40

File: src/node_jsrtapi.cc [hunk 1]
```diff
@@ -191,380 +269,389 @@ _Ret_maybenull_ JsValueRef CALLBACK CallbackWrapper(JsValueRef callee, bool isCo
...
-//Need to re-look as to create property descriptor correctly or not, cached property thing!!!
-void napi_set_function_name(napi_env e, napi_value func,
+napi_status napi_set_function_name(napi_env e, napi_value func,
   napi_propertyname name) {
-  JsErrorCode error = JsNoError;
-  JsValueRef object = reinterpret_cast<JsValueRef>(func);
-  JsPropertyIdRef propertyId = reinterpret_cast<JsPropertyIdRef>(name);
-  bool result;
-  error = JsDefineProperty(object, propertyId, nullptr, &result);
+  // TODO: Consider removing the napi_set_function_name API.
```
Comment (boingoing) [inline src/node_jsrtapi.cc:317]: +1
Opened #78 to track this.

File: src/node_jsrtapi.cc [hunk 2]
```diff
@@ -170,6 +170,84 @@ void CHAKRA_CALLBACK JsObjectWrapWrapperBeforeCollectCallback(JsRef ref, void* c
...
+napi_status napi_set_last_error(JsErrorCode jsError, void* engine_reserved = nullptr) {
+  napi_status status;
+  switch (jsError) {
+    case JsNoError: status = napi_ok; break;
+    case JsErrorNullArgument:
+    case JsErrorInvalidArgument: status = napi_invalid_arg; break;
+    case JsErrorPropertyNotString: status = napi_string_expected; break;
+    case JsErrorArgumentNotObject: status = napi_object_expected; break;
+    case JsErrorScriptException:
+    case JsErrorInExceptionState: status = napi_pending_exception; break;
```
Comment (boingoing) [inline src/node_jsrtapi.cc:241]: Does this mean we don't need NAPI methods to check for pending exception before returning? Seems to.

File: src/node_jsrtapi.cc [hunk 3]
```diff
@@ -191,380 +269,389 @@ _Ret_maybenull_ JsValueRef CALLBACK CallbackWrapper(JsValueRef callee, bool isCo
...
-    }
+  return napi_ok;
 }
 
-napi_value napi_get_undefined_(napi_env e) {
-  JsValueRef undefinedValue = nullptr;
-  JsErrorCode error = JsNoError;
-  error = JsGetUndefinedValue(&undefinedValue);
-  return reinterpret_cast<napi_value>(undefinedValue);
+napi_status napi_get_undefined_(napi_env e, napi_value* result) {
```
Comment (boingoing) [inline src/node_jsrtapi.cc:610]: This API has been renamed without the underscore in V8 version (napi_get_undefined). Should match here.

**Review by boingoing — APPROVED** [review]
Looks really good!

File: src/node_jsrtapi.cc (hunk 3 above)
Comment (jasongin) [reply src/node_jsrtapi.cc:610]: Fixed.

**Comment by jasongin** [comment]
Merged as f7a0d50142165369b4109269a834f44b923ac693.
Also I updated tests in the ChakraCore branch (copied from the V8 branch) as c4b425b8e8963375fc313c4069fabeea0aae76a1.

**Pull request closed without merging**